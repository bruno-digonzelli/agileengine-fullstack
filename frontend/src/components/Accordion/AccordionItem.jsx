import React, {useRef, useState, useLayoutEffect} from 'react'
import PropTypes from 'prop-types';

const AccordionItem = ({type, id, amount, effectiveDate}) => {
    // Hooks
    // - Ref
    const accordionContent = useRef(null);

    // - State
    const [active, setActive] = useState(false);
    const [height, setHeight] = useState("0px");

    // Handlers
    const handleAccordion = () => {
        setActive(prev => !prev);
    }

    useLayoutEffect(() => {
        setHeight(
            active ? `${accordionContent.current.scrollHeight}px` : '0px'
        );        
    }, [active])

    return ( 
        <div
            className={`accordion-item  ${active ? 'accordion-item--state-active' : ''} ${type === 'credit' ? 'accordion-item--type-credit' : 'accordion-item--type-debit'} p-4 mb-5`}
        >
            {/* Header */}
            <div className="accordion-item__header d-flex justify-content-between align-items-center">
                <div className="accordion-item__header-info d-flex justify-content-between pr-5">
                    <h2 className="accordion-item__header-title mr-3 text-capitalize">Type: {type}</h2>
                    <h3>Amount: {amount}</h3>
                </div>

                <button className="accordion-item__expand btn" onClick={handleAccordion}>{active ? 'Collapse' : 'Expand'}</button>
            </div>

            <div
                className="accordion-item__body overflow-hidden"
                ref={accordionContent}
                style={{maxHeight: `${height}`}}
            >
                <ul>
                    <li className="accordion-item__list mb-2">Transaction Id: {id}</li>
                    <li className="accordion-item__list mb-2">Type: {type}</li>
                    <li className="accordion-item__list mb-2">Amount: {amount}</li>
                    <li className="accordion-item__list">Effective Date: {effectiveDate}</li>
                </ul>
            </div>
        </div>
     );
}

AccordionItem.propTypes = {
    type: PropTypes.string.isRequired,
    id: PropTypes.string.isRequired,
    amount: PropTypes.number.isRequired,
    effectiveDate: PropTypes.string.isRequired
};

export default AccordionItem;