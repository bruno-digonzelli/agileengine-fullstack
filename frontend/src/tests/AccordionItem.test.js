import { render, screen, fireEvent } from '@testing-library/react';
import AccordionItem from '../components/Accordion/AccordionItem';

describe('AccordionItem', () => {
  const exampleItem = () => {
    const values = {
      type: 'credit',
      id: '844b85eb-1e79-4dd8-990c-b8ad7ec09213',
      amount: 13,
      effectiveDate: '2020-11-09T01:44:12.431Z'
    };

    return (
      render(<AccordionItem
        type={values.type}
        id={values.id}
        amount={values.amount}
        effectiveDate={values.effectiveDate}
      />)
    )
  };

  test('Accordion renders with values', () => {
    const accordion = exampleItem();

    expect(accordion).toMatchSnapshot();
  });

  test('Button changes text when clicked', () => {
    exampleItem();

    const buttonElement = screen.getByText('Expand');

    fireEvent.click(buttonElement);

    expect(buttonElement.innerHTML).toBe('Collapse');
  });
});
