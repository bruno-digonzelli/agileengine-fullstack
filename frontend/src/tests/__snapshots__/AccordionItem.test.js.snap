// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AccordionItem Accordion renders with values 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        class="accordion-item   accordion-item--type-credit p-4 mb-5"
      >
        <div
          class="accordion-item__header d-flex justify-content-between align-items-center"
        >
          <div
            class="accordion-item__header-info d-flex justify-content-between pr-5"
          >
            <h2
              class="accordion-item__header-title mr-3 text-capitalize"
            >
              Type: 
              credit
            </h2>
            <h3>
              Amount: 
              13
            </h3>
          </div>
          <button
            class="accordion-item__expand btn"
          >
            Expand
          </button>
        </div>
        <div
          class="accordion-item__body overflow-hidden"
          style="max-height: 0px;"
        >
          <ul>
            <li
              class="accordion-item__list mb-2"
            >
              Transaction Id: 
              844b85eb-1e79-4dd8-990c-b8ad7ec09213
            </li>
            <li
              class="accordion-item__list mb-2"
            >
              Type: 
              credit
            </li>
            <li
              class="accordion-item__list mb-2"
            >
              Amount: 
              13
            </li>
            <li
              class="accordion-item__list"
            >
              Effective Date: 
              2020-11-09T01:44:12.431Z
            </li>
          </ul>
        </div>
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      class="accordion-item   accordion-item--type-credit p-4 mb-5"
    >
      <div
        class="accordion-item__header d-flex justify-content-between align-items-center"
      >
        <div
          class="accordion-item__header-info d-flex justify-content-between pr-5"
        >
          <h2
            class="accordion-item__header-title mr-3 text-capitalize"
          >
            Type: 
            credit
          </h2>
          <h3>
            Amount: 
            13
          </h3>
        </div>
        <button
          class="accordion-item__expand btn"
        >
          Expand
        </button>
      </div>
      <div
        class="accordion-item__body overflow-hidden"
        style="max-height: 0px;"
      >
        <ul>
          <li
            class="accordion-item__list mb-2"
          >
            Transaction Id: 
            844b85eb-1e79-4dd8-990c-b8ad7ec09213
          </li>
          <li
            class="accordion-item__list mb-2"
          >
            Type: 
            credit
          </li>
          <li
            class="accordion-item__list mb-2"
          >
            Amount: 
            13
          </li>
          <li
            class="accordion-item__list"
          >
            Effective Date: 
            2020-11-09T01:44:12.431Z
          </li>
        </ul>
      </div>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
